README.md
# dMeasureQIM
[Quality Improvement Measures](https://www1.health.gov.au/internet/main/publishing.nsf/Content/PIP-QI_Incentive_guidance) module for dMeasure

## Structure of exported `CSV` (comma-separated-value) file

Based on [PIP QI Annotated Specifications (May 2020)](https://www1.health.gov.au/internet/main/publishing.nsf/Content/46506AF50A4824B6CA25848600113FFF/$File/PIP-QI-Annotated-Specifications.pdf), [PIP QI Technical Specifications V1.2](https://www1.health.gov.au/internet/main/publishing.nsf/Content/46506AF50A4824B6CA25848600113FFF/$File/PIP-QI-Technical-Specifications%20v1.2.pdf) and [PIP QI User Guide](https://www1.health.gov.au/internet/main/publishing.nsf/Content/46506AF50A4824B6CA25848600113FFF/$File/PIP%20QI%20-%20User%20Guide.pdf).

### Columns

1. `QIM` - `QIM 01` to `QIM 10`
  + `QIM 01` - Proportion of patients with diabetes with a current HbA1c result 
  + `QIM 02` - Proportion of patients whose smoking status has been recorded (a)/result (b)
  + `QIM 03` - Proportion of patients with a weight classification recorded (a)/result (b)
  + `QIM 04` - Proportion of patients aged 65 and over who were immunised against influenza
  + `QIM 05` - Proportion of patients with diabetes who were immunised against influenza
  + `QIM 06` - Proportion of patients with COPD who were immunised against influenza
  + `QIM 07` - Proportion of patients with an alcohol consumption status 
  + `QIM 08` - Proportion of patients with the necessary risk factors assessed to enable CVD assessment
  + `QIM 09` - Proportion of female patients with an up-to-date cervical screening
  + `QIM 10` - Proportion of patients with diabetes with a blood pressure result

2. `Age10` - Age group
  + `0` (0-4), `5` (5-14), `15` (15-24), `25` (25-34), `35` (35-44), `45` (45-54), `55` (55-64), `65` (65 and over *or* 65-69), `70` (70 to 74)
  + `QIM 01`, `QIM 05`, `QIM 10` - 8 groups (`0`, `5`, `15`, `25`, `35`, `45`, `55`, `65`)
  + `QIM 02`, `QIM 03`, `QIM 06`, `QIM 07` - 6 groups, 15 plus (`15`, `25`, `35`, `45`, `55`, `65`)
  + `QIM 04` - 65 plus (`65`) only
  + `QIM 08` - 3 groups, 45 plus (`45`, `55`, `65`) *plus* `35` for `Indigenous` state in (`Aboriginal`, `Torres Strait Islander`, `Both Aboriginal and Torres Strait Islander`)
  + `QIM 09` - 6 groups, 25 plus (`25`, `35`, `45`, `55`, `65`, `70`)

3. `Sex` - Person-sex, METeOR identifier 635126
  + `Male` - `1` Male
  + `Female` - `2` Female
  + `X` - `X` Indeterminate/Intersex/Unspecified
  + `NA` - not available (not specified)

4. `Indigenous` - Indigenous status, METeOR identifier 602543
  + `Aboriginal` - `1` 'Aboriginal but not Torres Strait Islander'
  + `Torres Strait Islander` - `2` 'Torres Strait Islander but not Aboriginal'
  + `Both Aboriginal and Torres Strait Islander` - `3` 'Both Aboriginal and Torres Strait Islander'
  + `Neither` - `4` 'Neither Aboriginal or Torres Strait Islander'
  + `Not stated` - `9` 'Not stated'
  
5. `Diabetes Type` - Diabetes status. METeOR identifier 270194. Only relevant to `QIM 01`, `QIM 05` and `QIM 10`. Left empty in all other QIM
  + `Type 1`
  + `Type 2`
  + `NA` - not available (not specified)
  
6. `Measure`
  + `QIM 01` - `HbA1C`
  + `QIM 02` - `Smoking`
  + `QIM 03` - `BMIClass`
  + `QIM 04` - `InfluenzaDone`
  + `QIM 05` - `InfluenzaDone`
  + `QIM 06` - `InfluenzaDone`
  + `QIM 07` - `AlcoholDone`
  + `QIM 08` - `CVDRiskDone`
  + `QIM 09` - `CSTDone`
  + `QIM 10` - `BPDone`
  
7. `State`
  + `QIM 01`, `QIM 04`, `QIM 05`, `QIM 06`, `QIM 07`, `QIM 08`, `QIM 10` - `TRUE` or `FALSE`
  + `QIM 02` - Smoking status : `Ex Smoker` (ex-smoker), `Non smoker` (never smoked), `Smoker` (current smoker) or `NA` (not available)
  + `QIM 03` - Body mass Index classification (METeOR identifier 270474) : `Healthy`, `Obese`, `Overweight`, `Underweight` or `NA` (not available)
  
8. `n` - the number of regular clients (three contacts - usually defined as three billed visits - over a two year period) who are in a group defined by columns 2, 3, 4, 5, 6 and 7 i.e. `Age10`, `Sex`, `Indigenous`, `Diabetes Type` (if applicable), `Measure` and `State`.
  + Supports small-cell-suppression - change `n` values less than 5 to 0 (zero). Section 2.4 of [PIP QI User Guide](https://www1.health.gov.au/internet/main/publishing.nsf/Content/46506AF50A4824B6CA25848600113FFF/$File/PIP%20QI%20-%20User%20Guide.pdf).

Additional columns of information which can either be extrapolatd from columns 1 to 8, or further define the nature of the information collection.

9. `ProportionDemographic` - the proportion of the demographic group (as defined by columns 2, 3, 4, 5 - (`Age10`, `Sex`, `Indigenous` and, if relevant, `Diabetes Type`) with the defined `State`

10. `DateFrom` and `DateTo` - the period (typically two years) of contacts

11. `ContactType` - the contact type, usually services i.e. billings

12. `MinContact` - the minimum number of contacts in the defined period, usually 3.

13. `Clinicians` - the clinicians who were contacted.

## Number of rows/groups

* QIM 01/05/10 : <img src = "https://render.githubusercontent.com/render/math?math=8*4*5*3*2"> each
* QIM 02 : <img src = "https://render.githubusercontent.com/render/math?math=6*4*5*4">
* QIM 03 : <img src = "https://render.githubusercontent.com/render/math?math=6*4*5*5">
* QIM 06/07 : <img src = "https://render.githubusercontent.com/render/math?math=6*4*5*2"> each 
* QIM 04 : <img src = "https://render.githubusercontent.com/render/math?math=1*4*5*2">
* QIM 08 : <img src = "https://render.githubusercontent.com/render/math?math=3*4*5*2%2B4*3*2">
* QIM 09 : allowing more than `Female` sex : <img src = "https://render.githubusercontent.com/render/math?math=6*4*5*2">

Total : <img src = "https://render.githubusercontent.com/render/math?math=960%2B480%2B600%2B40%2B960%2B240%2B240%2B(120%2B24)%2B240%2B960 = 4684">

## Structure of exported `JSON` (JavaScript Objection Notation) file

Based on [PIP QI Eligible Data Set - JSON Specification - version 1.1](https://www1.health.gov.au/internet/main/publishing.nsf/Content/46506AF50A4824B6CA25848600113FFF/$File/PIPQI-Eligible-Data-Set-JSON-Specification-v1.1.pdf)


NEWS.md
Changelog

# version 0.6.0
1st August 2021

## New

* `group_identification_suppression` option for `writeReportJSON`
  + suppress if numerator 2 or less OR difference between numerator and denominator is less than or equal to 2.
  + This is a simple probability-based disclosure suppression (l-diversity) against homogeneity attack,
    which overlaps with 'small cell suppression' of cells where the denominator is less than 5.

## Change

* `getReport` accepts `min_date` and `max_date` parameters
* compliance with [version 1.2 of PIP QI Improvement Measures - Technical Specifications](https://www1.health.gov.au/internet/main/publishing.nsf/Content/46506AF50A4824B6CA25848600113FFF/$File/PIP-QI-Technical-Specifications%20v1.2.pdf)
* QIM 04 (influenza 65+) `report_qim_65plus` - exclude patients who have been marked not to have an influenza immunization reminder
  + as per PIP QI Improvement Measures Technical Specifications V1.2 (22102020)
  + QIM report (called from `QIM_report_UI.R`) - do not include patients who have not visited within the last 15 months
  + as per PIP QI Improvement Measures Technical Specifications V1.2 (22102020)
* QIM 05 (influenza diabetes) `report_qim_diabetes`
  + if only 'influenza' measure chosen then exclude patients who have been marked not to have an influenza immunization reminder
  + if 'influenza' *and* another measure chosen, then if a patient has been marked as not to have an influenza immunization reminder,
    then the 'InfluenzaDone' value will be set to `NA`
  + as per PIP QI Improvement Meausres Technical Specifications V1.2 (22102020)
* QIM 06 (influenza COPD) `report_qim_copd` - exclude patients who have been marked not to have an influenza immunization reminder
  + as per PIP QI Improvement Measures Technical Specifications V1.2 (22102020)
* QIM 09 (cervical screening) `repot_qim_cost` - excludes patients who have been marked not have cervical screening or
  an indication that result may only be available elsewhere
  + as per PIP QI Improvement Measures Technical Specifications V1.2 (22102020)
  + Does not include if 'No longer requires cervical screening' is set
    OR 'Opt out of cervical screening' (reasons excluded include 'has screening at another clinic/elsewhere' and
    'refuses'  but does not include "Doesn't want reminders sent")

## Fix

* Add 'small cell suppression' and 'group identification suppression' to QIM 02a and QIM 03a

# version 0.5.0
12th January 2021

## New

* PIP QI Eligible Data set JSON specification version 1.1 export 

# version 0.4.1
16th September 2020

## Changes

* add 10-year cardiovascular (CVD) risk `frisk10` to QIM 08.
* increase maximum upload filesize to 300 megabytes

# version 0.4.0
18th August 2020

## New

* report creation download (comma-separated-values '.csv') and restore ('upload')
  + `getReport` to modify QIM-specific reports to generic 'long-format' which can be used by all QIM reports
  + support small number suppression
  + save and restore reports
  + charting
* compliance with version 1.1 of [PIP QI Improvement Measures - Technical Specifications v1.1 04052020](https://www1.health.gov.au/internet/main/publishing.nsf/Content/46506AF50A4824B6CA25848600113FFF/$File/PIP-QI-Technical-Specifications.pdf)
  + change `Age5` to `Age10` (minimum 0, maximum 65)
  + add `Indigenous`
  + `Sex` categories "Female", "Male", "X" (for indeterminate/intersec/unspecified) and "Not stated" (not available 'NA')
  + default demographic groups are `Age10`, `Sex`, `Indigenous`
  + for Smoking measure (QIM 02), `SmokingStatus` is used
  + for weight classification measure (QIM 03), `BMIclass` is used
    - now `BMIClass` is computed with age- and sex- specific ranges
  + for diabetes (QIM 01 and others), option to add diabetes type to description
    - `report_qim_diabetes` and others accepts a `type_diabetes` option
    - also determined  by active `self$qim_diabetes_showType` (and private and reactive versions)
    - userInterface adds pushbutton to turn on/off `self$qim_diabetes_showType` (default ON)
  + for CVD risk (QIM 08) add the condition that either diabetes must be diagnosed
    or has been checked (BSL or HbA1C) within the past two years
* `Proportion_Demographic` : proportion of measurement in each demographic sub-group
* `add_demographics` : function to add demographics to dataframe

## Changes

* add `store` option for `report_qim_*`, `list_qim_*` and `list_qim_*_appointments` methods.
  + by default, set to `TRUE`, store result to `self$qim_*_report`
  + if `FALSE`, the results are returned, but not stored to `self$qim_*_report`

## Bugfix

* QIM 02 15+ smoking `list_qim15_plus` : `ignoreOld` now ignores results more than one year old
* replace `dplyr::filter(x == max(x))` with `dplyr::arrange(dplyr::desc(x), .by_group = TRUE) %>>% dplyr::filter(dplyr::row_number() == 1)`
  + `arrange`/`row_number` breaks 'ties', where more than one of 'max' value
* 'Active' now not shown during 'Appointment' view (which is what was intended)

## Changes

* removal of `shinycssloaders`

# version 0.3.0
19th July 2020

* `dMeasureIntegration` for auto-loading module

# version 0.2.0
17th July 2020

* move DailyMeasure/GPstat webUI from `DailyMeasure` to `dMeasureQIM`
  + adds `shinydashboardmenuItem`, `dMeasureShinytabItems`,
    `datatableUI`, `datatableServer` in `userInterface.R`

# version 0.1.6
10th July 2020

* `WeightDone` in `report_qim_15plus` depends on `BMIDate` instead of `WeightDate`

# version 0.1.5
16th May 2020

* add `max_date` to
  + `report_qim_active`,
  + `list_qim_*`, `list_qim_*_appointments`, `report_qim_*`

# version 0.1.4
3rd May 2020

* add `max_date` to call to `list_contact_count`

# version 0.1.3
10th April 2020

* rintrojs walkthrough (introduction) `steps_introduction_df`

# version 0.1.2
19th February 2020

* change in QIM_15plus

R/QIM_report.R
#' @name qim_report
#' @title dMeasure Quality Improvement Measures - reports for QIM
#'
#' @include QualityImprovementMeasures.R
NULL

#' getReport
#'
#' @md
#'
#' @description create reports in a 'standard' format
#'   from the 10 QIM measures
#'
#' gets report from 'report_function'
#' then adds extra 'DiabetesType' column if necessary
#' adds more columns describing the measurement
#' renames the 'result' column to 'State'
#'
#' small cell suppression supported
#'   section 2.4 'Small cell suppression' of
#'   "Practice Incentives Program Quality Improvement Measures User Guide"
#'   https://www1.health.gov.au/internet/main/publishing.nsf/Content/
#'   46506AF50A4824B6CA25848600113FFF/$File/PIP%20QI%20-%20User%20Guide.pdf
#'
#' @param report_function the dMQIM method to call
#' @param date_from start date
#' @param date_to end date
#' @param contact_type types of contacts e.g. Services,
#' @param min_contact minimum number of contacts to be 'valid'
#' @param min_date most recent contact must be at least min_date
#' @param max_date most recent contact at most max_date
#' @param progress object for displaying progress. Leave as `NA` if
#'   not required
#' @param progress_detail message to report in the 'progress' window
#' @param measure some (but not all) `report_function` can
#'   choose the 'measure' if not required, then use `NA`
#' @param require_type_diabetes diabetes methods may give option
#'   to specify diabetes types
#' @param qim_name the name of the QIM e.g. 'QIM 01'
#' @param measure_name the name of the measure e.g. 'HbA1C'
#' @param state_variable_name the name of the column in the return
#'   from `report_function` which becomes the 'State' column
#' @param small_cell_suppression 'suppress' (return zero, or remove
#'   row entirely) if count is less than 5.
#' @param include_all_demographic_groups include all possible
#'   demographic groups/subgroups, even if there was no count
#'
#' @export
getReport <- function(
  report_function,
  date_from, date_to,
  contact_type, min_contact,
  min_date = as.Date("2000-01-01"),
  max_date = Sys.Date(),
  # no effective limitations on min/max date by default
  progress = NA,
  progress_detail,
  measure = NA,
  require_type_diabetes,
  qim_name, measure_name,
  state_variable_name,
  small_cell_suppression = FALSE,
  include_all_demographic_groups = TRUE
) {
  if (is.environment(progress)) { # a OOP object (as opposed to 'NA')
    progress$inc(amount = 1, detail = progress_detail)
  }

  function_args <- list(
    contact = TRUE, date_from = date_from, date_to = date_to,
    min_contact = min_contact,
    # 'default' is clinician list chosen in right panel
    min_date = min_date, max_date = max_date,
    contact_type = contact_type,
    demographic = c("Age10", "Sex", "Indigenous"),
    ignoreOld = TRUE, lazy = FALSE, store = FALSE
  )
  if (!is.na(measure)) {
    function_args <- append(
      function_args,
      list(measure = measure)
    )
  }
  if (require_type_diabetes) {
    function_args <- append(
      function_args,
      list(type_diabetes = TRUE)
    )
  }
  qim_report <- do.call(report_function, function_args) %>>%
    dplyr::mutate(
      QIM = qim_name, Measure = measure_name
    ) %>>% {
      if (require_type_diabetes) {
        . # 'DiabetesType' should already exit
      } else {
        dplyr::mutate(., DiabetesType = "")
        # create a 'blank'
      }
    } %>>%
    dplyr::rename(
      State = !!state_variable_name,
      ProportionDemographic = Proportion_Demographic
    ) %>>% {
      if (small_cell_suppression) {
        dplyr::filter(., n >= 5)
        # 'suppress' the cell (removed completely!) if < 5
      } else {
        .
      }
    } %>>%
    dplyr::select(
      QIM, Age10, Sex, Indigenous, DiabetesType,
      Measure, State, n, ProportionDemographic
    )
  # keep Age10, Sex, Indigenous, DiabetesType, HbA1CDone,
  # n, Proportion_Demographic

  if (include_all_demographic_groups) {
    qim_report <- dMeasureQIM::fill_demographics(qim_report)
  }

  return(qim_report)
}


#' writeReportJSON
#'
#' @md
#'
#' @description write reports to PIP QI Eligible Data Set JSON Specification 1.1
#'
#' @param d data
#' @param date_to chosen end date
#' @param author_id identifier from source software of the authenticated user who
#'   initiated the extract
#' @param practice_id an identifier for the practice
#' @param product the name of the software that produced the file
#' @param report Must be "PIP QI"
#' @param version version number of the specification used to generate the JSON
#' @param small_cell_suppression 'suppress' (return NA, or remove
#'   row entirely) if count is denominator less than 5
#'
#'   This measure is explictly allowed under '[Practice Incentives Program Quality Improvement Incentives Quality Improvement Measures User Guide for General Practices (2020)](https://www1.health.gov.au/internet/main/publishing.nsf/Content/46506AF50A4824B6CA25848600113FFF/$File/PIP-QI-User-Guide-Practices.pdf)',
#'   Section 2.7 'Is the data de-identified?', page 8.
#'
#' @param group_identification_suppression numerator 2 or less
#'   OR difference between numerator and denominator is less than or equal to 2.
#'   This is a simple probability-based disclosure suppression
#'   (l-diversity) against homogeneity attack e.g. ALL or NONE of the
#'   patients in a sub-group have a specified state.
#'
#'   This measure is not explicitly mentioned under the PIP QIM User Guide, but
#'   is consistent with avoiding disclosure of information about individual
#'   patients.
#' @param indigenous_aggregate simplify indigenous groups to
#'   INDIGENOUS, NON-INDIGENOUS, NOT STATED
#' @param sex_aggregate simplify sex groups to
#'   MALE, FEMALE, INDETERMINATE/INTERSEX/UNSPECIFIED/NOT STATED/INADEQUATELY DESCRIBED
#'
#' @return JSON string
#'
#' @export
writeReportJSON <- function(
  d, date_to,
  author_id = "bpsrawdata",
  practice_id,
  product = "GPstat",
  report = "PIP QI",
  version = "1.2",
  small_cell_suppression = TRUE,
  group_identification_suppression = FALSE,
  indigenous_aggregate = TRUE, sex_aggregate = TRUE) {

  options(digits.secs = 3) # print second component to 3 decimal places
  x <- paste(date_to, "22:59:59.999") # one hour before midnight
  attr(x, "tzone") <- "UTC" # convert from local timezone to UTC
  extraction_datetime <- as.character(x)

  d <- d %>>%
    dplyr::filter(DateTo == date_to) %>>% # just keep the required date information
    dplyr::select(QIM, Age10, Sex, Indigenous, DiabetesType, State, n)

  if (indigenous_aggregate) {
    # INDIGENOUS, NON-INDIGENSOUS or 'Not stated'
    d <- d %>>%
      dplyr::mutate(
        Indigenous = dplyr::if_else(
          Indigenous %in% c("Aboriginal", "Both Aboriginal and Torres Strait Islander", "Torres Strait Islander"),
          "INDIGENOUS", # all indigenous groups combined to 'INDIGENOUS'
          Indigenous
        )
      ) %>>%
      dplyr::mutate(
        Indigenous = dplyr::if_else(
          Indigenous == "Neither",
          "NON-INDIGENOUS", # re-code 'not indigenous'
          Indigenous
        )
      ) %>>%
      dplyr::group_by(QIM, Age10, Sex, Indigenous, DiabetesType, State) %>>%
      dplyr::summarize(n = sum(n)) %>>% # aggregate indigenous classifications
      dplyr::ungroup()
  }

  if (sex_aggregate) {
    d <- d %>>%
      dplyr::mutate(
        Sex = dplyr::if_else(
          Sex %in% c("Not stated", "X"), # these two are combined
          "INDETERMINATE/INTERSEX/UNSPECIFIED/NOT STATED/INADEQUATELY DESCRIBED",
          Sex
        )
      ) %>>%
      dplyr::group_by(QIM, Age10, Sex, Indigenous, DiabetesType, State) %>>%
      dplyr::summarize(n = sum(n)) %>>% # aggregate sex classifications
      dplyr::ungroup()
  }

  if (small_cell_suppression | group_identification_suppression) {
    # 'suppress' small groups to zero. helps with de-identification
    suppress_small_cells <- function(d) {
      return(
        d %>>%
          dplyr::mutate(
            numerator = dplyr::if_else(
              denominator != 0 & denominator < 5 & small_cell_suppression,
              # potentially identifying information if group size < 5
              # avoid 'suppression' if the denominator is zero!
              as.integer(NA),
              as.integer(numerator)
            )
          ) %>>%
          dplyr::mutate(
            numerator = dplyr::if_else(
              denominator != 0 & (numerator <= 2 | (denominator - numerator) <= 2) & group_identification_suppression,
              # numerator is 2 or less, or very close to the denominator
              # this is a type of probability-based disclosure suppression
              as.integer(NA),
              as.integer(numerator)
            )
          )
      )
    }
  } else {
    suppress_small_cells <- function(d) {
      # no suppression, just return the current dataframe
      return(d)
    }
  }

  d_denominators <- d %>>%
    dplyr::group_by(QIM, Age10, Sex, Indigenous, DiabetesType) %>>%
    dplyr::summarise(denominator = sum(n)) %>>% # the total in each sub-group
    dplyr::ungroup()

  d <- d %>>%
    dplyr::left_join(
      d_denominators,
      by = c("QIM", "Age10", "Sex", "Indigenous", "DiabetesType")
    ) %>>% # add denominators to each group
    dplyr::filter(State != "FALSE") %>>% # these are excluded from the report!
    dplyr::rename(sex = Sex, age_group = Age10,
                  indigenous_status = Indigenous, diabetes_type = DiabetesType,
                  numerator = n) %>>%
    dplyr::mutate(sex = toupper(sex),
                  indigenous_status = toupper(indigenous_status))

  qim_df <- list(
    extraction_datetime = jsonlite::unbox(extraction_datetime),
    author_id = jsonlite::unbox(author_id),
    practice_id = jsonlite::unbox(practice_id),
    product = jsonlite::unbox(product),
    report = jsonlite::unbox(report),
    version = jsonlite::unbox(version)
  )
  # all these single elements are 'unboxed'
  # an alternative is to define as dataframe
  #  in which case all the 'qim' names below will need to be
  #  defined as lists instead of dataframes
  #  (appending '%>>% list()' is enough)

  qim_measures <- unique(d %>>% dplyr::pull(QIM))
  # the measures included in the report

  if ("QIM 01" %in% qim_measures) {
    qim_df[["qim01-diabetes"]] <- d %>>%
      dplyr::filter(QIM == "QIM 01") %>>%
      dplyr::mutate(hba1c_result = "RECORDED") %>>%
      dplyr::select(sex, age_group, indigenous_status, diabetes_type,
                    hba1c_result, numerator, denominator) %>>%
      suppress_small_cells()
  }

  if ("QIM 02" %in% qim_measures) {
    qim_df[["qim02a-smoking"]] <- d %>>%
      dplyr::filter(QIM == "QIM 02") %>>%
      dplyr::mutate(smoking_status = "RECORDED") %>>%
      dplyr::select(sex, age_group, indigenous_status,
                    smoking_status, State, numerator, denominator) %>>%
      dplyr::mutate(
        State = dplyr::if_else(
          State %in% c("Ex smoker", "Non smoker", "Smoker"), # these three are combined
          TRUE,
          NA
        )
      ) %>>%
      dplyr::filter(!is.na(State)) %>>%
      dplyr::group_by(sex, age_group, indigenous_status,
                      smoking_status, State, denominator) %>>%
      dplyr::summarize(numerator = sum(numerator)) %>>%
      dplyr::ungroup() %>>%
      dplyr::select(sex, age_group, indigenous_status,
                    smoking_status, numerator, denominator) %>>%
      suppress_small_cells()

    d_subset <- d %>>%
      dplyr::filter(QIM == "QIM 02") %>>%
      dplyr::select(sex, age_group, indigenous_status,
                    smoking_status = State, numerator, denominator) %>>%
      dplyr::filter(smoking_status %in% c("Ex smoker", "Non smoker", "Smoker")) %>>%
      dplyr::mutate(
        smoking_status = dplyr::case_when(
          smoking_status == "Ex smoker" ~ "EX",
          smoking_status == "Non smoker" ~ "NEVER",
          smoking_status == "Smoker" ~ "CURRENT"
        )
      )

    qim_df[["qim02b-smoking"]] <- d_subset %>>%
      dplyr::select(sex, age_group, indigenous_status,
                    smoking_status, numerator) %>>%
      # re-calculate the denominator (does not include the NA 'not recorded')
      dplyr::left_join(
        d_subset %>>%
          dplyr::group_by(sex, age_group, indigenous_status) %>>%
          dplyr::summarize(denominator = sum(numerator)) %>>%
          dplyr::ungroup() %>>%
          dplyr::select(sex, age_group, indigenous_status, denominator),
        by = c("sex", "age_group", "indigenous_status")
      ) %>>%
      suppress_small_cells()
  }

  if ("QIM 03" %in% qim_measures) {
    qim_df[["qim03a-bmi"]] <- d %>>%
      dplyr::filter(QIM == "QIM 03") %>>%
      dplyr::mutate(bmi = "RECORDED") %>>%
      dplyr::select(sex, age_group, indigenous_status,
                    bmi, State, numerator, denominator) %>>%
      dplyr::mutate(
        State = dplyr::if_else(
          State %in% c("Healthy", "Obese", "Overweight", "Underweight"), # these three are combined
          TRUE,
          NA
        )
      ) %>>%
      dplyr::filter(!is.na(State)) %>>%
      dplyr::group_by(sex, age_group, indigenous_status,
                      bmi, State, denominator) %>>%
      dplyr::summarize(numerator = sum(numerator)) %>>%
      dplyr::ungroup() %>>%
      dplyr::select(sex, age_group, indigenous_status,
                    bmi, numerator, denominator) %>>%
      suppress_small_cells()

    d_subset <- d %>>%
      dplyr::filter(QIM == "QIM 03") %>>%
      dplyr::select(sex, age_group, indigenous_status,
                    bmi = State, numerator, denominator) %>>%
      dplyr::filter(bmi %in% c("Healthy", "Obese", "Overweight", "Underweight")) %>>%
      dplyr::mutate(bmi = toupper(bmi))

    qim_df[["qim03b-bmi"]] <- d_subset %>>%
      dplyr::select(sex, age_group, indigenous_status,
                    bmi, numerator) %>>%
      # re-calculate the denominator (does not include the NA 'not recorded')
      dplyr::left_join(
        d_subset %>>%
          dplyr::group_by(sex, age_group, indigenous_status) %>>%
          dplyr::summarize(denominator = sum(numerator)) %>>%
          dplyr::ungroup() %>>%
          dplyr::select(sex, age_group, indigenous_status, denominator),
        by = c("sex", "age_group", "indigenous_status")
      ) %>>%
      suppress_small_cells()
  }

  if ("QIM 04" %in% qim_measures) {
    qim_df[["qim04-influenza_65years"]] <- d %>>%
      dplyr::filter(QIM == "QIM 04") %>>%
      dplyr::mutate(influenza_immun_status = "IMMUNISED") %>>%
      dplyr::select(sex, age_group, indigenous_status,
                    influenza_immun_status, numerator, denominator) %>>%
      suppress_small_cells()
  }

  if ("QIM 05" %in% qim_measures) {

    d_subset <- d %>>%
      dplyr::filter(QIM == "QIM 05") %>>%
      dplyr::mutate(influenza_immun_status = "IMMUNISED") %>>%
      dplyr::select(sex, age_group, indigenous_status, diabetes_type,
                    influenza_immun_status, numerator, denominator)

    qim_df[["qim05-influenza_diabetes"]] <- d_subset %>>%
      dplyr::group_by(sex, age_group, indigenous_status,
                      influenza_immun_status) %>>%
      # this QIM does not actually have diabetes sub-groups
      dplyr::summarize(numerator = sum(numerator)) %>>%
      dplyr::ungroup() %>>%
      dplyr::left_join(
        d_subset %>>%
          dplyr::group_by(sex, age_group, indigenous_status) %>>%
          dplyr::summarize(denominator = sum(denominator)) %>>%
          # denominator includes those who are not immunized
          dplyr::ungroup(),
        by = c("sex", "age_group", "indigenous_status")
      ) %>>%
      dplyr::select(sex, age_group, indigenous_status,
                    influenza_immun_status, numerator, denominator) %>>%
      suppress_small_cells()
  }

  if ("QIM 06" %in% qim_measures) {
    qim_df[["qim06-influenza_copd"]] <- d %>>%
      dplyr::filter(QIM == "QIM 06") %>>%
      dplyr::mutate(influenza_immun_status = "IMMUNISED") %>>%
      dplyr::select(sex, age_group, indigenous_status,
                    influenza_immun_status, numerator, denominator) %>>%
      suppress_small_cells()
  }

  if ("QIM 07" %in% qim_measures) {
    qim_df[["qim07-alcohol"]] <- d %>>%
      dplyr::filter(QIM == "QIM 07") %>>%
      dplyr::mutate(alcohol_status = "RECORDED") %>>%
      dplyr::select(sex, age_group, indigenous_status,
                    alcohol_status, numerator, denominator) %>>%
      suppress_small_cells()
  }

  if ("QIM 08" %in% qim_measures) {
    qim_df[["qim08-cvd"]] <- d %>>%
      dplyr::filter(QIM == "QIM 08") %>>%
      dplyr::mutate(cvd_risk = "RECORDED") %>>%
      dplyr::select(sex, age_group, indigenous_status,
                    cvd_risk, numerator, denominator) %>>%
      suppress_small_cells()
  }

  if ("QIM 09" %in% qim_measures) {
    qim_df[["qim09-cervical"]] <- d %>>%
      dplyr::filter(QIM == "QIM 09") %>>%
      dplyr::mutate(cervical_screen_status = "SCREENED") %>>%
      dplyr::select(sex, age_group, indigenous_status,
                    cervical_screen_status, numerator, denominator) %>>%
      suppress_small_cells()
  }

  if ("QIM 10" %in% qim_measures) {

    d_subset <- d %>>%
      dplyr::filter(QIM == "QIM 10") %>>%
      dplyr::mutate(bp_result = "RECORDED") %>>%
      dplyr::select(sex, age_group, indigenous_status, diabetes_type,
                    bp_result, numerator, denominator)

    qim_df[["qim10-diabetes_bp"]] <- d_subset %>>%
      dplyr::group_by(sex, age_group, indigenous_status,
                      bp_result) %>>%
      # this QIM does not actually have diabetes sub-groups
      dplyr::summarize(numerator = sum(numerator)) %>>%
      dplyr::ungroup() %>>%
      dplyr::left_join(
        d_subset %>>%
          dplyr::group_by(sex, age_group, indigenous_status) %>>%
          dplyr::summarize(denominator = sum(denominator)) %>>%
          # denominator includes those who are not immunized
          dplyr::ungroup(),
        by = c("sex", "age_group", "indigenous_status")
      ) %>>%
      dplyr::select(sex, age_group, indigenous_status,
                    bp_result, numerator, denominator) %>>%
      suppress_small_cells()
  }

  return(jsonlite::toJSON(qim_df, na = "string", pretty = TRUE))
  # 'NA' values have been small cell suppressed
}
